pkg/kubevirt/utils.go

```diff
@@ -138,36 +146,86 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template, cloudInitType)
+	switch cloudInitType {
+	case "CloudInitNoCloud":
+		err := mergo.Merge(&template.Spec.Volumes[index], cloudInitVolume)
+		return nil, err
+	case "CloudInitConfigDrive":
+		template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
+	}
+	if found, _ := detectCloudInitDisk(template); !found {
+		template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+	}
+
+	return template, nil
+}
+
+// nodeRole returns the role of this node ("control-plane" or "worker").
+func nodeRole(ctx *context.MachineContext) string {
+	if util.IsControlPlaneMachine(ctx.Machine) {
+		return constants.ControlPlaneNodeRoleValue
+	}
+	return constants.WorkerNodeRoleValue
+}
+
+func cloudinitVolume(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, ctx *context.MachineContext, preferredType string) kubevirtv1.Volume {
+	switch preferredType {
+	case "CloudInitNoCloud":
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
 				},
 			},
-		},
+		}
+	default:
+		return kubevirtv1.Volume{
+			Name: cloudInitVolumeName,
+			VolumeSource: kubevirtv1.VolumeSource{
+				CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
+					UserDataSecretRef: &corev1.LocalObjectReference{
+						Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+					},
+				},
+			},
+		}
 	}
-	template.Spec.Volumes = append(template.Spec.Volumes, cloudInitVolume)
+}
 
-	cloudInitDisk := kubevirtv1.Disk{
+func cloudinitDisk(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec, preferredType string) kubevirtv1.Disk {
+	return kubevirtv1.Disk{
 		Name: cloudInitVolumeName,
 		DiskDevice: kubevirtv1.DiskDevice{
 			Disk: &kubevirtv1.DiskTarget{
 				Bus: "virtio",
 			},
 		},
 	}
-	template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+}
 
-	return template
+func detectCloudInitVolumeType(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec) (preferredType string, index int) {
+	preferredType = "CloudInitConfigDrive"
+	for i, v := range vmi.Spec.Volumes {
+		if v.CloudInitNoCloud != nil && v.Name == cloudInitVolumeName {
+			preferredType, index = "CloudInitNoCloud", i
+			break
+		}
+	}
+	return preferredType, index
 }
```

Comment:
Personally, I don't like named return values and I find them confusing. If there is no reason to use them, please consider to drop them.

Also, I think we could write a shorter, simpler version of this function. How do you feel about this suggestion?

```suggestion
func detectCloudInitVolumeType(vmi *kubevirtv1.VirtualMachineInstanceTemplateSpec) (string, int) {
	for i, v := range vmi.Spec.Volumes {
		if v.CloudInitNoCloud != nil && v.Name == cloudInitVolumeName {
			return "CloudInitNoCloud", i
		}
	}
	return "CloudInitConfigDrive", 0
}
```